components/FlexibleHeader/tests/unit/FlexibleHeaderTopSafeAreaTests.m

```diff
@@ -0,0 +1,314 @@
+      [[FlexibleHeaderTopSafeAreaTestsFakeViewController alloc] init];
+  _topSafeArea.topSafeAreaSourceViewController = viewController;
+
+  // When
+  viewController.topSafeAreaInset = 44;
+  [_topSafeArea safeAreaInsetsDidChange];
+
+  // Then
+  XCTAssertEqualWithAccuracy(_topSafeArea.topSafeAreaInset, 44, 0.0001);
+}
+
+// This test simulates a non-notch device hiding its status bar in reaction to the flexible header
+// shifting off-screen. In this case the top safe area inset should stay "pinned" to the previous
+// known value (of 20).
+- (void)testTopSafeAreaInsetDoesNotChangeWhileStatusBarIsShiftedWithNonNotchSafeAreaSize {
+  // Given
+  _topSafeArea.inferTopSafeAreaInsetFromViewController = YES;
+  _topSafeArea.delegate = _delegate;
+  FlexibleHeaderTopSafeAreaTestsFakeViewController *viewController =
+      [[FlexibleHeaderTopSafeAreaTestsFakeViewController alloc] init];
+  viewController.topSafeAreaInset = 20;
+  _topSafeArea.topSafeAreaSourceViewController = viewController;
+
+  // When
+  _delegate.isStatusBarShifted = YES;
+  viewController.topSafeAreaInset = 0;
+  [_topSafeArea safeAreaInsetsDidChange];
+
+  // Then
+  XCTAssertEqualWithAccuracy(_topSafeArea.topSafeAreaInset, 20, 0.0001);
+}
+
+- (void)testTopSafeAreaInsetDoesChangeWhileStatusBarIsShiftedWithAnyOtherSize {
+  // Given
+  _topSafeArea.inferTopSafeAreaInsetFromViewController = YES;
+  _topSafeArea.delegate = _delegate;
+  FlexibleHeaderTopSafeAreaTestsFakeViewController *viewController =
+      [[FlexibleHeaderTopSafeAreaTestsFakeViewController alloc] init];
+  viewController.topSafeAreaInset = 44;
+  _topSafeArea.topSafeAreaSourceViewController = viewController;
+
+  // When
+  _delegate.isStatusBarShifted = YES;
+  viewController.topSafeAreaInset = 0;
+  [_topSafeArea safeAreaInsetsDidChange];
+
+  // Then
+  XCTAssertEqualWithAccuracy(_topSafeArea.topSafeAreaInset, 0, 0.0001);
+}
+
+- (void)testTopSafeAreaInsetResetsToZeroWhenViewControllerIsSetToNil {
+  // Given
+  _topSafeArea.inferTopSafeAreaInsetFromViewController = YES;
+  FlexibleHeaderTopSafeAreaTestsFakeViewController *viewController =
+      [[FlexibleHeaderTopSafeAreaTestsFakeViewController alloc] init];
+  viewController.topSafeAreaInset = 44;
+  _topSafeArea.topSafeAreaSourceViewController = viewController;
+
+  // When
+  _topSafeArea.topSafeAreaSourceViewController = nil;
+
+  // Then
+  XCTAssertEqualWithAccuracy(_topSafeArea.topSafeAreaInset, 0, 0.0001);
+}
+
+- (void)testTopSafeAreaInsetResetsToViewControllerWhenViewControllerIsNilThenReSet {
+  // Given
+  _topSafeArea.inferTopSafeAreaInsetFromViewController = YES;
+  FlexibleHeaderTopSafeAreaTestsFakeViewController *viewController =
+      [[FlexibleHeaderTopSafeAreaTestsFakeViewController alloc] init];
+  viewController.topSafeAreaInset = 44;
+  _topSafeArea.topSafeAreaSourceViewController = viewController;
+
+  // When
+  _topSafeArea.topSafeAreaSourceViewController = nil;
```

Comment:
Let's change the `topAreaInset` value in the VC before re-setting it to ensure that it detects the current value and not the original value.